### PR TITLE
 feat: Changes to UnitPerf measuring

### PR DESCRIPTION
**Pull request by elliotfreebairn1**
* Resolves: # 
* Target version: master 

### Summary
Working on issue #9276

Signed-off-by: elliotfreebairn <elliot.freebairn@collabora.com>
Change-Id: I868a0a1c91b30392ab002ce149cefae0dd3036cc

### TODO

- [ ] ...

### Checklist

- [ ] I have run `make prettier-write` and formatted the code.
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by elliotfreebairn1** [comment]
This is not fully finished as i haven't created a impress and draw trace but i wanted to see if i'm along the right path especially on if the hook is in the right place and if i messed up anywhere in here. 
Here is screenshot of what the outcome of a unit test would look like:

File: test/UnitPerf.cpp [hunk 1]
```diff
@@ -86,13 +94,29 @@ void UnitPerf::invokeWSDTest()
...
+
+    std::cerr << "test: " << cpuTime << "us\n";
 
     exitTest(TestResult::Ok);
 }
 
+//Called when document loading process starts e.g. setup finishes
+void UnitPerf::onDocumentLoading()
+{
+    stats->perfSetupLoadEditPhases("Setup");
```
Comment (mmeeks) [inline test/UnitPerf.cpp:111]: Hmm; sounds like stats->endPhase(foo) might be a better name (?) I assume that ends it (?)

File: tools/Replay.hpp [hunk 2]
```diff
@@ -216,32 +267,104 @@ struct Stats {
         dumpMap(_recvd);
     }
 
-    void dumpStressToCSV(size_t runMs)
+
+    //loading phase will record the set up times and loaded phase records the document loading times
```
Comment (mmeeks) [inline tools/Replay.hpp:271]: //<space>text =)

File: tools/Replay.hpp [hunk 3]
```diff
@@ -216,32 +267,104 @@ struct Stats {
...
+    void perfSetupLoadEditPhases(std::string phase)
```
Comment (mmeeks) [inline tools/Replay.hpp:272]: So - I guess, it's perhaps a better self documenting API to have an enumeration of phase type; there are some nice macros Ash has that allow you to get a nice string name from each Enum value - we can then use as the string for that; and then it's more clear what the phases are and that we move through them; checkout: common/Log.hpp and the STATE_ENUM macro =)

File: tools/Replay.hpp [hunk 4]
```diff
@@ -216,32 +267,104 @@ struct Stats {
...
+        _start = std::chrono::steady_clock::now();
+
+        size_t cpuTime = _timer->elapsedTime().count();
+        _timer.reset(new Util::SysStopwatch);
 
-        if (file.tellp() == 0)
+        perfStatsList.push_back(GetStressStats(runMs, phase));
+        perfStatsList.push_back(GetCPUUSageStats(cpuTime, phase));
+
+        if(phase == "Edit")
```
Comment (mmeeks) [inline tools/Replay.hpp:284]: enums more reliable for comparison;

File: tools/Replay.hpp [hunk 5]
```diff
@@ -216,32 +267,104 @@ struct Stats {
...
+        PerfMetricInfo cpuStats = {testPhase, "CPU Usage (us)", cpuUsage};
+        return cpuStats;
+    }
+
+    std::vector<PerfMetricInfo> GetNetworkStats(size_t recievedKb, size_t sentKb, std::string testPhase)
+    {
+        std::vector<PerfMetricInfo> networkStatsList;
+
+        networkStatsList.push_back(PerfMetricInfo{testPhase, "Bytes recieved (kB)",  recievedKb});
+        networkStatsList.push_back(PerfMetricInfo{testPhase,"Bytes sent (kB)", sentKb});
```
Comment (mmeeks) [inline tools/Replay.hpp:327]: one space after comma, not none, not two ;-)
Please read your diff carefully before committing.

File: tools/Replay.hpp [hunk 6]
```diff
@@ -216,32 +267,104 @@ struct Stats {
...
+    }
+
+    void dumpPerfStatsToCSV(std::vector<PerfMetricInfo> perfData)
+    {
+        std::ofstream file("PerformanceMetricsSummary.csv", std::ios::out | std::ios::app);
 
         if(file.tellp() == 0)
         {
-            file << "Incoming bandwidth (kB/s)" << ",Outgoing bandwidth (kB/s)";
+            file << "Commit Hash" << "," << "Date" << "," << "Test" << "," << "Phase" << "," << "Metric" << "," << "Value";
```
Comment (mmeeks) [inline tools/Replay.hpp:344]: All Capitalized Except value =)

File: wsd/DocumentBroker.cpp [hunk 7]
```diff
@@ -873,6 +873,11 @@ bool DocumentBroker::download(
     {
         _docState.setStatus(DocumentState::Status::Downloading);
 
+        if(_unitWsd != nullptr)
+        {
+            _unitWsd->onDocumentLoading();
```
Comment (mmeeks) [inline wsd/DocumentBroker.cpp:878]: Looks reasonable I guess =)

**Review by mmeeks — COMMENTED** [review]
Looking reasonable =) needs a bit of fixup, some rebasing / squashing fixes into commits and so on - but getting somewhere good :-)

File: test/UnitPerf.cpp (hunk 1 above)
Comment (elliotfreebairn1) [reply test/UnitPerf.cpp:111]: Yeah that would make sense, this gets called when the document is getting downloaded from storage (wopi host), so i assume from the beginning of the unit test to then is the setup phase?

File: tools/Replay.hpp (hunk 3 above)
Comment (elliotfreebairn1) [reply tools/Replay.hpp]: I've tried to implement this and it does work but i'm unsure if this is the correct way to implement the Enum value?

Comment (mmeeks) [reply tools/Replay.hpp]: Looks lovely =)

File: tools/Replay.hpp [hunk 8]
```diff
@@ -113,11 +119,13 @@ struct Histogram {
             }
         }
 
-        file << totalTiles << ",";
-        file << subTenCount << ",";
-        file << subOneHundredCount << ",";
-        file << overOneHundredCount << ",";
-        file << "\n";
+        std::vector<PerfMetricInfo> latencyStatsList;
+        latencyStatsList.push_back(PerfMetricInfo {testPhase,typeOfLatency + " Total tiles", totalTiles});
```
Comment (mmeeks) [inline tools/Replay.hpp:123]: consistency 'Phase,typeOf' - no space - =)

File: tools/Replay.hpp [hunk 9]
```diff
@@ -29,6 +29,20 @@
 
 #include <iostream>
 #include <fstream>
+#include <config_version.h>
+#include <ctime>
+#include <Util.hpp>
+#include <common/Log.hpp>
+
+
```
Comment (mmeeks) [inline tools/Replay.hpp:37]: Unusually large to have 3x spaces - vertical space is a non-renewable resource to be used tastefully =)

File: tools/Replay.hpp [hunk 10]
```diff
@@ -152,6 +165,8 @@ struct Stats {
     std::unordered_map<std::string, MessageStat> _recvd;
     std::unordered_map<std::string, MessageStat> _sent;
 
+    std::vector<PerfMetricInfo> perfStatsList;
```
Comment (mmeeks) [inline tools/Replay.hpp:168]: Is that a member - if so it needs an _ prefix so code-readers can tell that -cf. all other member variables =)

File: tools/Replay.hpp [hunk 11]
```diff
@@ -256,46 +268,106 @@ struct Stats {
...
+    }
+
+    void dumpPerfStatsToCSV(std::vector<PerfMetricInfo> perfData)
+    {
+        std::ofstream file("PerformanceMetricsSummary.csv", std::ios::out | std::ios::app);
 
         if(file.tellp() == 0)
         {
-            file << "Bytes recieved (kB)" << ", Bytes sent(kB)";
+            file << "COMMIT HASH" << "," << "DATE" << "," << "TEST" << "," << "PHASE" << "," << "METRIC" << "," << "value";
```
Comment (mmeeks) [inline tools/Replay.hpp:346]: Hmm - now all of these legends are ALLCAPS - shouting at me, but the last item is still lower-case "value" - why ? =) I was suggesting that if you have: "Foo, Baa, baz" - something is wrong with baz - instead we got "FOO, BAA, baz" ;-) Can we have: "Foo, Baa, Baz" ?

**Review by mmeeks — APPROVED** [review]
Better; it can go in without further review; just trying to build your sub-clinical neurosis & code skimming speed with a few whitespace comments that jarred as I read =)

**Comment by elliotfreebairn1** [comment]
Thank you for picking them out, i definitely need to switch on my brain when it comes to code skimming. Will fix these then merge :)

**Pull request merged**